Integrate SwanLab for offline/online experiment tracking and local visualization

ShaohonChen opened the pull request:
# What does this PR do?

This PR introduces [SwanLab](https://github.com/SwanHubX/SwanLab), a lightweight open-source experiment tracking tool, as a new logging option for the training framework. The integration provides both online and offline tracking capabilities, along with a local dashboard for visualizing results.

SwanLab has previously supported tracking the Transformers training framework through external callbacks (find more information [here](https://docs.swanlab.cn/guide_cloud/integration/integration-huggingface-transformers.html)), serving a wide range of users—especially those in regions with limited network connectivity, such as China. With this official integration, we aim to further enhance the developer experience by making tracking more seamless and user-friendly.

![image](https://github.com/user-attachments/assets/767696e0-355f-4d7b-a462-f3237c9dde5c)

**Additional information about this PR** is a detailed overview of the changes and usage instructions.

<img width="1585" alt="swanlab-terminal" src="https://github.com/user-attachments/assets/c68d4b9b-14a6-4785-8c2c-32578c5af2a1" />
<img width="1702" alt="swanlab-web" src="https://github.com/user-attachments/assets/3fa774bd-c133-4792-bfd1-74059ec3cb9b" />


## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? (I don't see any tests for any of the callbacks but please let me know if I missed them somewhere. )

## Who can review?

@SunMarc  I have reviewed recent merges related to training tracking, and it seems that you are the most relevant reviewer for this PR. Could you please help review it or forward it to the appropriate person? Thank you!

---

## Additional information about this PR

### Key Features of SwanLab Integration

**1. Online and Offline Tracking:**

- **Online Mode**: Track experiments remotely and store data on SwanLab's cloud platform.
- **Offline Mode**: Use a local dashboard to visualize training logs without an internet connection.

**2. Hardware Monitoring:**

- Automatically tracks GPU usage, power consumption, temperature, and other hardware metrics.
- Supports NVIDIA GPUs and Huawei Ascend NPUs.

**3. Remote Access:**

- View training progress remotely via the SwanLab web interface or mobile app.

**4. Local Dashboard:**

- Includes an open-source local dashboard for offline visualization of training logs.

### Usage guidline

**Step 0: Set Up code and environment**

Following the **transformers** official text classification example:

```bash
# prepare code and environments
git clone https://github.com/huggingface/transformers
cd transformers
pip install -e .
cd examples/pytorch/text-classification
pip install -r requirements.txt
# 
```

**Step 1: Set Up SwanLab Online Tracking**

Install:

```bash
pip install swanlab
```

To use SwanLab's online tracking, log in to the [SwanLab website](https://swanlab.cn/) and obtain your API key from the [Settings page](https://swanlab.cn/space/~/settings). Then, authenticate using the following command:

```bash
swanlab login
```

If you prefer offline mode, skip this step.

**Step 2: Configure SwanLab as the Logger and run example**

To enable SwanLab as the experiment tracker, add `--use_swanlab` to your training command. For example, using the  workflow:

```bash
python run_glue.py \
  --model_name_or_path google-bert/bert-base-cased \
  --dataset_name imdb  \
  --do_train \
  --do_predict \
  --max_seq_length 128 \
  --per_device_train_batch_size 32 \
  --learning_rate 2e-5 \
  --num_train_epochs 3 \
  --output_dir /tmp/imdb/
```

<img width="1585" alt="swanlab-terminal" src="https://github.com/user-attachments/assets/414acf52-fddd-4cab-9a9a-7e41c626df7d" />

[visualization demo](https://swanlab.cn/@ShaohonChen/text-classification/runs/rpkiz27hi8vd86t15egzn/chart)

<img width="1702" alt="swanlab-web" src="https://github.com/user-attachments/assets/7098ae5e-9e66-4628-a13a-e6ec50509e04" />

If you want to use local tracking, you simply set environment variable following:

```bash
# Linux & Mac
export SWANLAB_MODE = "local"
# Win
$env:SWANLAB_MODE = "local"
```

Alternatively, you can configure SwanLab using environment variables:

```bash
export SWANLAB_API_KEY=<your_api_key>          # Set API key for online tracking
export SWANLAB_LOG_DIR=<local_log_path>        # Set local log directory
export SWANLAB_MODE=<mode>                    # Set tracking mode: cloud (default), cloud-only, local, or disabled
```

**Step 3: View Training Logs**

After logging in, you will see a confirmation message:

- **Online Tracking:**  View logs on the [SwanLab website](https://swanlab.cn/).

<img width="1702" alt="swanlab-web" src="https://github.com/user-attachments/assets/3fa774bd-c133-4792-bfd1-74059ec3cb9b" />

<img width="1693" alt="swanlab-config" src="https://github.com/user-attachments/assets/570ff004-3b6b-4bbc-aa5e-b5c2f2991d13" />

<img width="1688" alt="swanlab-gpu" src="https://github.com/user-attachments/assets/13746ad7-0396-4f06-ab67-b0728b14bc65" />

For more details, refer to the [SwanLab Cloud Documentation](https://docs.swanlab.cn/en/guide_cloud/experiment_track/view-result.html).

- **Offline Tracking**: Use the local dashboard to visualize logs:

[here](https://docs.swanlab.cn/en/guide_cloud/self_host/remote-view.html#access-the-experiment-dashboard-from-the-local-machine) for more information.

```bash
swanlab watch
```

<img width="766" alt="swanlab-local-terminal" src="https://github.com/user-attachments/assets/aea84972-3b37-41b2-83b2-13409b04a38e" />

<img width="1705" alt="swanlab-local-web" src="https://github.com/user-attachments/assets/219becec-679b-4122-b535-80270083b97f" />


For advanced configurations, such as setting a custom port, refer to the [Offline Dashboard Documentation](https://docs.swanlab.cn/guide_cloud/self_host/offline-board.html) and [CLI Documentation](https://docs.swanlab.cn/api/cli-swanlab-watch.html#%E8%AE%BE%E7%BD%AEip%E5%92%8C%E7%AB%AF%E5%8F%A3%E5%8F%B7).

ShaohonChen committed:
add swanlab integration

ShaohonChen committed:
feat(integrate): add SwanLab as an optional experiment tracking tool in transformers

- Integrated SwanLab into the transformers library as an alternative for experiment tracking.
- Users can now log training metrics, hyperparameters, and other experiment details to SwanLab by setting `report_to="swanlab"` in the `TrainingArguments`.
- Added necessary dependencies and documentation for SwanLab integration.

ShaohonChen committed:
Fix the spelling error of SwanLabCallback in callback.md

ShaohonChen committed:
Merge branch 'main' into integrate-swanlab

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+    """
+
+    def __init__(self):
+        has_swanlab = is_swanlab_available()
+        if not has_swanlab:
+            raise RuntimeError("SwanLabCallback requires swanlab to be installed. Run `pip install swanlab`.")
+        if has_swanlab:
+            import swanlab
+
+            self._swanlab = swanlab
```
```suggestion
        if not is_swanlab_available():
            raise RuntimeError("SwanLabCallback requires swanlab to be installed. Run `pip install swanlab`.")
        import swanlab
        
        self._swanlab = swanlab
```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+                - If the user is already logged in, this environment variable takes precedence over locally stored
+                login information.
+
+        - **SWANLAB_PROJECT** (`str`, *optional*, defaults to `None`):
+            Set this to a custom string to store results in a different project. If not specified, the name of the running
+            directory is used.
+
```
maybe typo ? 
```suggestion
        - **SWANLAB_PROJECTS** (`str`, *optional*, defaults to `None`):
            Set this to a custom string to store results in a different project. If not specified, the name of the running
            directory is used.

```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+            API address for the SwanLab cloud environment for private version (its free)
+
+        """
+        if self._swanlab is None:
+            return
```
Will never be none 
```suggestion
```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+            # add config parameters (run may have been created manually)
+            self._swanlab.config.update(combined_dict)
+
+            # add number of model parameters to wandb config
```
```suggestion
            # add number of model parameters to swanlab config
```

ShaohonChen replied:
Oh dear, I didn't notice that I forgot to update the comments when using the Weights & Biases callback! Thanks for your review!

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+                modelcard.AUTOGENERATED_TRAINER_COMMENT += f"\n{badge_markdown}"
+
+    def on_train_begin(self, args, state, control, model=None, **kwargs):
+        if self._swanlab is None:
+            return
```
```suggestion
```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+            "total_flos",
+        ]
+
+        if self._swanlab is None:
+            return
```
```suggestion
```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+            )
+
+    def on_predict(self, args, state, control, metrics, **kwargs):
+        if self._swanlab is None:
+            return
```
```suggestion
```

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,173 @@ def on_train_end(self, args, state, control, **kwargs):
+
+    def on_save(self, args, state, control, **kwargs):
+        if self._log_model is not None and self._initialized and state.is_world_process_zero:
+            logger.warning(
+                "SwanLab does not currently support the save mode functionality. "
+                "This feature will be available in a future release.",
+                UserWarning,
+                stacklevel=2,
+            )
```
Let's keep it simple
```suggestion
            logger.warning(
                "SwanLab does not currently support the save mode functionality. "
                "This feature will be available in a future release."
            )
```

SunMarc reviewed (commented):
Thanks for adding this ! Just a few nits !

ShaohonChen commented:
Thanks for the review! I'll update it shortly.

ShaohonChen committed:
Apply suggestions from code review

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

ShaohonChen committed:
Fix typo in comment

ShaohonChen committed:
Fix typo in comment

ShaohonChen reviewed (commented):
I have made the required changes and fixed the typo. Thanks to @SunMarc for the review!

ShaohonChen committed:
Merge branch 'integrate-swanlab' of https://github.com/ShaohonChen/transformers into integrate-swanlab

ShaohonChen commented:
I ran the example in examples/pytorch/text-classification with the updated code. Everything looks good!
Here is a screenshot and the link to the test run:

<img width="1733" alt="Screenshot 2025-03-01 at 01 35 52" src="https://github.com/user-attachments/assets/6deeb6cc-321b-4130-bd2b-4c924757e4cc" />

<img width="1959" alt="Screenshot 2025-03-01 at 01 36 33" src="https://github.com/user-attachments/assets/c26eb9ef-7270-4eb6-bcfb-d35e148fc84f" />

Really appreciate your suggestions. Please have a look when you get a chance!

SunMarc reviewed (approved):
Thanks for iterating !

SunMarc commented on src/transformers/integrations/integration_utils.py:
```diff
@@ -2141,6 +2147,161 @@ def on_train_end(self, args, state, control, **kwargs):
+                - If the user is already logged in, this environment variable takes precedence over locally stored
+                login information.
+
+        - **SWANLAB_PROJECT** (`str`, *optional*, defaults to `None`):
```
the typo is still there

ShaohonChen replied:
Oh wow, this might actually be a translation mistake in our official docs. I'll go through all the comments in this PR to double-check the spelling and fix any issues. Thanks for the review!

AAssets replied:
I have standardized the naming of SWANLAB_PROJECT, changed 'working directory' to the more accurate 'running directory,' and refined the LOG_DIR comment for better clarity. The typos should now be resolved. Thank you for reviewing the changes!

ShaohonChen replied:
I noticed the mistake in the use of the SWANLAB_PROJECT variable later in the code! Thanks for helping to correct it. It looks like the automated checks are still having some issues. I'll continue working on fixing them.

AAssets committed:
Fix typos and update comments

SAKURA-CAT commented:
It seems the test timed out. Could it be due to excessive computational load? I haven't found any explicit error messages.

Zeyi-Lin committed:
fix annotation

ShaohonChen committed:
Merge pull request #1 from Zeyi-Lin/integrate-swanlab

Fix: annotation

SAKURA-CAT committed:
chore: opt some comments

ShaohonChen committed:
Merge pull request #2 from SAKURA-CAT/integrate-swanlab

chore: opt some comments

ShaohonChen committed:
Merge branch 'main' into integrate-swanlab

SAKURA-CAT commented:
> It seems the test timed out. Could it be due to excessive computational load?

Re-running the test succeeded. 🤔

ShaohonChen commented:
@SunMarc I have updated the code and fixed the incorrect comments. After merging the latest main branch, I noticed that the automated tests seem to have failed. I checked, and it doesn't appear to be an issue caused by my changes. Could you help me restart the tests?

ShaohonChen committed:
Merge branch 'main' into integrate-swanlab

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36433). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ShaohonChen committed:
Merge branch 'main' into integrate-swanlab

SunMarc committed:
Merge branch 'main' into integrate-swanlab

ShaohonChen commented:
Oops, my bad. I accidentally hit the review button.

SunMarc commented:
Keep up the good work !
